Could not compile (implicit declaration of function '__v_thread_wait' is invalid)

<!-- It is advisable to update all relevant modules using `v outdated` and `v install` -->
**V doctor:**
```
OS: macos, macOS, 11.6,
Processor: 16 cpus, 64bit, little endian
CC version: Apple clang version 13.0.0 (clang-1300.0.29.3)

getwd: /Users/usr/code/lang/v/vlang
vmodules: /Users/usr/.vmodules
vroot: /Users/usr/src/v
vexe: /Users/usr/src/v/v
vexe mtime: 2021-09-29 02:35:36
is vroot writable: true
is vmodules writable: true
V full version: V 0.2.4 bd0653a.8dde9d4

Git version: git version 2.33.0
Git vroot status: weekly.2021.39-25-g8dde9d4a
.git/config present: true
thirdparty/tcc status: thirdparty-macos-amd64 689c8a02

```

**What did you do?**
`v -g -o vdbg cmd/v && vdbg study/issue/dup_def_thread_wait.v`
```v

fn main() {
	sa := SA{}
	sb := SA{}

	sa.wait()
	sb.wait()
}
struct SA {
mut:
	threads []thread
}
pub fn (self SA) wait() {
	self.threads.wait()
}
struct SB {
mut:
	threads []thread
}
pub fn (self SB) wait() {
	self.threads.wait()
}

```

**What did you expect to see?**

no error


**What did you see instead?**
```
==================
/tmp/v/dup_def_thread_wait.11190777266797634777.tmp.c:1822:3: error: implicit declaration of function '__v_thread_wait' is invalid in C99 [-Werror,-Wimplicit-function-declaration]
                __v_thread_wait(t);
                ^
1 error generated.
...
==================
(Use `v -cg` to print the entire error message)

builder error: 
==================
C error. This should never happen.

This is a compiler bug, please report it using `v bug file.v`.

https://github.com/vlang/v/issues/new/choose

You can also use #help on Discord: https://discord.gg/vlang

```

--- a/vlib/v/gen/c/cgen.v
+++ b/vlib/v/gen/c/cgen.v
@@ -887,6 +887,23 @@ fn (mut g Gen) write_shareds() {
 	}
 }
 
+fn (mut g Gen) register_thread_void_wait_call() {
+	if '__v_thread_wait' !in g.waiter_fns {
+		g.gowrappers.writeln('void __v_thread_wait(__v_thread thread) {')
+		if g.pref.os == .windows {
+			g.gowrappers.writeln('\tu32 stat = WaitForSingleObject(thread, INFINITE);')
+		} else {
+			g.gowrappers.writeln('\tint stat = pthread_join(thread, (void **)NULL);')
+		}
+		g.gowrappers.writeln('\tif (stat != 0) { _v_panic(_SLIT("unable to join thread")); }')
+		if g.pref.os == .windows {
+			g.gowrappers.writeln('\tCloseHandle(thread);')
+		}
+		g.gowrappers.writeln('}')
+		g.waiter_fns << '__v_thread_wait'
+	}
+}
+
 fn (mut g Gen) register_thread_array_wait_call(eltyp string) string {
 	is_void := eltyp == 'void'
 	thread_typ := if is_void { '__v_thread' } else { '__v_thread_$eltyp' }
@@ -896,6 +913,7 @@ fn (mut g Gen) register_thread_array_wait_call(eltyp string) string {
 	if fn_name !in g.waiter_fns {
 		g.waiter_fns << fn_name
 		if is_void {
+			g.register_thread_void_wait_call()
 			g.gowrappers.writeln('
 void ${fn_name}($thread_arr_typ a) {
 	for (int i = 0; i < a.len; ++i) {

--- a/vlib/v/parser/parse_type.v
+++ b/vlib/v/parser/parse_type.v
@@ -70,6 +70,9 @@ pub fn (mut p Parser) parse_array_type() ast.Type {
 		// error is set in parse_type
 		return 0
 	}
+	if elem_type.idx() == ast.thread_type_idx {
+		p.register_auto_import('sync.threads')
+	}
 	mut nr_dims := 1
 	// detect attr
 	not_attr := p.peek_tok.kind != .name && p.peek_token(2).kind !in [.semicolon, .rsbr]

--- a/vlib/v/tests/go_array_wait_without_go_test.v
+++ b/vlib/v/tests/go_array_wait_without_go_test.v
@@ -1,0 +1,27 @@
+fn test_go_array_wait_without_go() {
+	sa := SA{}
+	sb := SA{}
+
+	sa.wait()
+	sb.wait()
+
+	assert true
+}
+
+struct SA {
+mut:
+	threads []thread
+}
+
+pub fn (self SA) wait() {
+	self.threads.wait()
+}
+
+struct SB {
+mut:
+	threads []thread
+}
+
+pub fn (self SB) wait() {
+	self.threads.wait()
+}

